### PR TITLE
Wizard for creating linkml2sheets templates

### PR DESCRIPTION
**Pull request by turbomam**
Will revert to draft after discussing with CJM 2022-07-15

No related issue?

This branch also contains some makefile rules illustrating round tripping from schema to sheets and back

### CODE REVIEW

File: schemasheets/template_wizard.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+# todo: work on YAML-serialized columns like annotation and alt_descriptions
+# todo: style guide, esp use of external terms and consistent grammar
+
+# todo add customized first header BEFORE the unfiltered sheet is written
+
+# todo: group and or describe mappings
+# todo: comments, notes, todos distinctions
+
+# todo: generate third row for internal separators and annotation names
+
```
Comment (turbomam) [inline schemasheets/template_wizard.py:29]: use python dataclasses instead of init methods

**Review by cmungall — CHANGES_REQUESTED** [review]
can you remove all unneccessary files from the commit?

**Comment by turbomam** [comment]
> can you remove all unnecessary files from the commit?

Thanks for looking at this again.

Shoot, I thought this was one of my tidier PRs. Can you give an example of what I should remove?

**Comment by turbomam** [comment]
I wasn't planning on using this for the schemasheets part of the workshop, if that makes any difference in the priority.

File: schemasheets/list_element_slots.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+slots = meta_view.class_induced_slots(meta_element)
+
+slot_names = [str(slot.alias) for slot in slots]
+
+slot_names.sort()
+
+as_if_tsv = "\t".join(slot_names)
+
+print(as_if_tsv)
```
Comment (cmungall) [inline schemasheets/list_element_slots.py:29]: if this is intended as a module that code can import, there should be no side effects

If it's not a module, then remove it.

If it is useful for testing or demo purposes, make it a test or a notebook

File: template_wizard.Makefile [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  		
+# target/output/schema.yaml: target/output/slot.tsv
+#	poetry run sheets2linkml target/output/*.tsv > $@
+
+target/output/key_counts.csv:
+	poetry run python schemasheets/list_element_slots.py
+
+# template_wizard_clean
+
+target/output/NMDC_schema_slot_class_merged_with_filtered_rels.tsv:
```
Comment (cmungall) [inline template_wizard.Makefile:30]: this doesn't look like something that belongs in a general purpose schemasheets repo?

File: schemasheets/template_wizard.py (hunk 1 above)
Comment (cmungall) [reply schemasheets/template_wizard.py:46]: I think you can resolve this comment, looks lovely with dataclasses!

File: schemasheets/template_wizard.py [hunk 4]
```diff
@@ -0,0 +1,1011 @@
...
+            element="slot_definition", selected_schema="meta"
+        )
+        class_slots_slotnames = list(set(class_names).union(set(slot_names)))
+        class_slots_slotnames.sort()
+        current_slot_attributes = self.return_slot_attributes(class_slots_slotnames)
+        self.class_slots_slotnames = class_slots_slotnames
+        self.slot_attributes = current_slot_attributes
+
+    def get_proj_elem_usage(self):
+        # todo I'm not convinced that this is getting all frequently used keys
```
Comment (cmungall) [inline schemasheets/template_wizard.py:119]: if not convinced the best thing to do is write a test!

File: schemasheets/template_wizard.py [hunk 5]
```diff
@@ -0,0 +1,1011 @@
...
+        Dict[SLOT_NAME, Dict[METAMODEL_SLOT_NAME, SLOT_VALUE]]
+    ] = None
+
+    # todo refactor this? dict with name and count had been handy for writing to CSV
+    #  but were not doing that anymore
+    proj_elem_usage: Optional[List[Dict[PROJECT_SLOT_NAME, Any]]] = None
+
+    class_slots_slotnames: Optional[List[PROJECT_SLOT_NAME]] = None
+
+    def get_element_annotations(self) -> List:
```
Comment (cmungall) [inline schemasheets/template_wizard.py:68]: make sure each public methods has a docstring

use _ if not public

Comment (cmungall) [inline schemasheets/template_wizard.py:68]: the type is incomplete - list of what?

File: schemasheets/template_wizard.py [hunk 6]
```diff
@@ -0,0 +1,1011 @@
...
+        """
+        iteratively collects all element annotations
+        """
+        elements = self.project_view.all_elements()
+        annotations = set()
+        for ek, ev in elements.items():
+            if ev.annotations:
```
Comment (cmungall) [inline schemasheets/template_wizard.py:75]: minor: unnecessary, the way pythongen works (and meta.py is made by pythongen) is such that every list
is prepopulated as an empty list

**Comment by joshmoore** [comment]
Some feedback:
 - Firstly, thanks! 💐  This is precisely what I was looking for today.
 - Copying the template_wizard.makefile into a new directory was a solid starting point
   - I uncommented the slot.tsv and schema.yaml targets. They additionally need dependencies on all the other tasks.
   - I replaced the ndmc.yaml links with `$(SCHEMA)` and all references to ndmc with `$(PREFIX)` which worked well.
   - I replaced `biosample` with my target class.
 - `project_source` in list_element_slots.py needs configuring and the use of "work" needs replacing with "target" (for this Makefile)
 - A few code changes were needed:
   - schema_exporter.py line 171 needs a check for a SlotDefinition (for my model)
   - A few places don't seem to detect all the internal columns (e.g. `'attributes', 'class_uri', 'description', 'class', 'from_schema', 'mixins'`)
   - schemamaker.py line 145 is getting a `dict` rather than a `list` in my case. This is a weird one.
 - Round-tripping failed with `ValueError: Enter an interpretation for attributes` but since I primarily wanted to get to a TSV (or actually, a google sheet) this was more than enough. Thanks again.

File: template_wizard.Makefile (hunk 3 above)
Comment (joshmoore) [reply template_wizard.Makefile:30]: Happy to submit variables to replace the specific if desired.

File: schemasheets/list_element_slots.py (hunk 2 above)
Comment (joshmoore) [reply schemasheets/list_element_slots.py:29]: My experience with the Makefile was that it's required which likely speaks towards having it as a command.

**Comment by sierra-moxon** [comment]
@joshmoore @turbomam - do you think we can close this PR and the associated ticket?  Or instead, should we get it to a point where we can merge it into the repository? :)

**Comment by joshmoore** [comment]
I discussed briefly with @cmungall in February and I think there was another path forward that was preferred. But in general, big :+1: from my sides to having some form of command or recipe to create the necessary pieces here.

**Pull request closed without merging**